# Hadoop: Map-Reduce 

ПРИМЕЧАНИЕ: для этого блокнота вам следует запустить сервер с Hadoop (с YARN) и средой Spark.

[Библиотека программного обеспечения Apache Hadoop](https://hadoop.apache.org/) — это фреймворк, который позволяет выполнять распределенную обработку больших наборов данных в кластерах компьютеров с использованием простых моделей программирования. Он разработан для масштабирования от отдельных серверов до тысяч машин, каждая из которых предлагает локальные вычисления и хранение. Вместо того, чтобы полагаться на оборудование для обеспечения высокой доступности, сама библиотека разработана для обнаружения и обработки сбоев на уровне приложений, тем самым предоставляя высокодоступный сервис поверх кластера компьютеров, каждый из которых может быть подвержен сбоям.

Текущая установка включает следующие модули из экосистемы Hadoop:

- __Hadoop Common:__ Общие утилиты, которые поддерживают другие модули Hadoop.
- __Hadoop Distributed File System (HDFS™):__ Распределенная файловая система, которая обеспечивает высокопроизводительный доступ к данным приложений.
- __Hadoop YARN:__ Фреймворк для планирования заданий и управления ресурсами кластера.
- __Hadoop MapReduce:__ Система на основе YARN для параллельной обработки больших наборов данных.
- __Spark™:__ Быстрый и универсальный вычислительный движок для данных Hadoop. Spark предоставляет простую и выразительную модель программирования, которая поддерживает широкий спектр приложений, включая ETL, машинное обучение, потоковую обработку и вычисление графов.

Псевдораспределенный режим (кластер с одним узлом)

Hadoop также может быть запущен на одном узле в [псевдораспределенном режиме](https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-common/SingleCluster.html#Pseudo-Distributed_Operation), где каждый демон Hadoop работает в отдельном процессе Java.

В псевдораспределенном режиме мы также используем только один узел, но главное, что кластер моделируется, что означает, что все процессы внутри кластера будут работать независимо друг от друга. Все демоны, которые являются Namenode, Datanode, Secondary Name node, Resource Manager, Node Manager и т. д., будут работать как отдельный процесс на отдельной JVM (виртуальной машине Java) или, можно сказать, работать на разных процессах Java, поэтому он называется псевдораспределенным.

<b>ОЧЕНЬ ВАЖНОЕ ПРИМЕЧАНИЕ: экземпляр Hadoop, установленный в среде «Hadoop (с YARN) и Spark», был разработан только для образовательных целей и НЕ СОХРАНЯЕТ ДАННЫЕ после остановки сервера. Вы можете создавать или удалять файлы в файловой системе HDFS, записывать данные во время сеанса, но при следующем запуске сервера Jupyter будет чистая файловая система без данных.</b>

In [1]:
import os
import re
import json
import socket
import subprocess
import pandas as pd
from pyspark import SparkContext, SparkConf
from pyspark.sql import SparkSession
from pyspark.sql import utils as pu
from pyspark.sql import functions as F
from pyspark.sql import types as pt

In [2]:
YARN_PORT = 8088

# директория по умолчанию для пользователя `jovyan`
WORK_DIR = '/jovyan'

In [3]:
def hdfs_dirs(path, filter_str=''):
    """
    Возвращает файлы по указанному пути в виде списка. 
    Имена файлов можно фильтровать по параметру `filter_str`,
    например, `filter_str='csv'` отобразит только файлы `csv`.
    
    """
    process = subprocess.Popen(
        ['hdfs', 'dfs', '-ls', path], 
        stdout=subprocess.PIPE, 
        stderr=subprocess.PIPE
    )
    out, err = process.communicate()
    dirs = out.decode('utf-8').split('\n')
    dirs = list(filter(lambda x: filter_str in x, dirs))
    dirs = list(map(lambda x: x.split(' ')[-1], dirs))
    return dirs

def file_content(path):
    """
    Возвращает содержимое файла.
    Аналогично команде `cat`.
    
    """
    process = subprocess.Popen(
        ['hdfs', 'dfs', '-cat', path], 
        stdout=subprocess.PIPE, 
        stderr=subprocess.PIPE
    )
    out, err = process.communicate()
    return out.decode('unicode_escape')

In [4]:
!hdfs dfs -ls /

Found 4 items
drwxr-xr-x   - hadoop supergroup           0 2025-06-30 04:37 /hbase
drwxr-xr-x   - jovyan hadoopusers          0 2025-06-30 04:37 /jovyan
drwxrwxrwx   - hadoop supergroup           0 2025-06-30 04:37 /tmp
drwxr-xr-x   - jovyan hadoopusers          0 2025-06-30 04:37 /user


Посмотрим, что лежит в рабочей директории

In [5]:
!hdfs dfs -ls {WORK_DIR}

Здесь пусто. 


Скопируем файл с логами студентов на образовательном портале. 

In [11]:
!hdfs dfs -put aggrigation_logs_per_week.csv {WORK_DIR}

In [12]:
!hdfs dfs -ls {WORK_DIR}

Found 1 items
-rw-r--r--   1 jovyan hadoopusers   39030677 2025-06-30 04:39 /jovyan/aggrigation_logs_per_week.csv


Вызовем фанкцию вывода списка файлов

In [13]:
hdfs_dirs(WORK_DIR, 'csv')

['/jovyan/aggrigation_logs_per_week.csv']

Выведем первые 100 символов. 

In [15]:
file_content ('/jovyan/aggrigation_logs_per_week.csv')[:100]

'courseid,userid,num_week,s_all,s_all_avg,s_course_viewed,s_course_viewed_avg,s_q_attempt_viewed,s_q_'

# Команда hdfs fsck

hdfs fsck — это утилита для проверки целостности файлов и блоков в HDFS. Она позволяет выявить проблемы, такие как отсутствующие блоки, поврежденные данные или другие несоответствия.

Аргументы: 
- files 
вывести информацию о файле, включая его размер, количество блоков и другие метаданные.

- blocks
добавляет информацию о блоках, из которых состоит файл. В HDFS файлы разбиваются на блоки (обычно размером 128 МБ или 256 МБ), которые распределяются по узлам кластера. Этот флаг покажет, сколько блоков занимает файл, их размеры и статус.

- racks
добавляет информацию о том, на каких стойках (racks) расположены блоки файла. В HDFS стойки используются для обеспечения отказоустойчивости и оптимизации производительности. Этот флаг покажет, как блоки распределены по стойкам в кластере.

In [16]:
%%bash
hdfs fsck /jovyan/aggrigation_logs_per_week.csv -files -blocks -racks

Connecting to namenode via http://0.0.0.0:9870/fsck?ugi=jovyan&files=1&blocks=1&racks=1&path=%2Fjovyan%2Faggrigation_logs_per_week.csv


FSCK started by jovyan (auth:SIMPLE) from /10.112.133.96 for path /jovyan/aggrigation_logs_per_week.csv at Mon Jun 30 04:39:34 UTC 2025

/jovyan/aggrigation_logs_per_week.csv 39030677 bytes, replicated: replication=1, 1 block(s):  OK
0. BP-743194254-10.112.133.96-1751258213750:blk_1073741837_1013 len=39030677 Live_repl=1  [/default-rack/10.112.133.96:9866]


Status: HEALTHY
 Number of data-nodes:	1
 Number of racks:		1
 Total dirs:			0
 Total symlinks:		0

Replicated Blocks:
 Total size:	39030677 B
 Total files:	1
 Total blocks (validated):	1 (avg. block size 39030677 B)
 Minimally replicated blocks:	1 (100.0 %)
 Over-replicated blocks:	0 (0.0 %)
 Under-replicated blocks:	0 (0.0 %)
 Mis-replicated blocks:		0 (0.0 %)
 Default replication factor:	1
 Average block replication:	1.0
 Missing blocks:		0
 Corrupt blocks:		0
 Missing replicas:		0 (0.0 %)

Erasure Coded Block Groups:
 Total size:	0 B
 Total files:	0
 Total block groups (validated):	0
 Minimally erasure-coded block groups:	0
 Ov

# Пояснения к выводу
Connecting to namenode via http://0.0.0.0:9870/fsck?ugi=jovyan&files=1&blocks=1&racks=1&path=%2Fjovyan%2Faggrigation_logs_per_week.csv
FSCK started by jovyan (auth:SIMPLE) from /10.112.129.148 for path /jovyan/aggrigation_logs_per_week.csv at Mon Jan 27 11:39:52 UTC 2025

/jovyan/aggrigation_logs_per_week.csv 39030677 bytes, replicated: replication=1, 1 block(s):  OK

#### Размер файла: 39 030 677 байт (~37 МБ).
#### Репликация: replication=1 (файл хранится только в одном экземпляре).
#### Число блоков: 1 (весь файл помещён в один блок).
#### Статус: OK (файл здоров).

<i> 0. BP-433375171-10.112.129.148-1737977684764:blk_1073741837_1013 len=39030677 Live_repl=1  [/default-rack/10.112.129.148:9866]  </i>
#### Информация о блоке:
#### Идентификатор блока: blk_1073741837_1013.
#### Длина блока: len=39030677 (совпадает с размером файла).
#### Количество активных реплик: Live_repl=1 (блок доступен только на одном узле).
#### Расположение блока: /default-rack/10.112.129.148:9866 (узел на rack'е default-rack).

Status: HEALTHY
 Number of data-nodes:	1
 Number of racks:		1
 Total dirs:			0
 Total symlinks:		0

#### HDFS работает с одним узлом хранения данных.
#### Rack'ов (групп узлов) тоже один.



# Команда hdfs dfs -setrep

команда HDFS, которая изменяет фактор репликации (количество копий) для указанного файла или директории.

Пример использования:
```
%%bash
hdfs dfs -setrep -w 2 /jovyan/aggrigation_logs_per_week.csv
```
Аргументы команды:
- -w
Этот флаг указывает, что команда должна ожидать завершения процесса репликации. Без этого флага команда просто запустит процесс репликации в фоновом режиме.

- 2
Это новое значение фактора репликации. В данном случае файл будет иметь 2 копии (реплики) в HDFS.

- /jovyan/aggrigation_logs_per_week.csv
Это путь к файлу в HDFS, для которого изменяется фактор репликации.


В нашей среде данная команда не сработает, поскольку у нас в наличие только одна DataNode.

## Map-reduce

Hadoop MapReduce — это программная среда для простого написания приложений, которые обрабатывают огромные объемы данных (многотерабайтные наборы данных) параллельно на больших кластерах (тысячи узлов) стандартного оборудования надежным и отказоустойчивым способом.

Задание MapReduce обычно разбивает входной набор данных на независимые фрагменты, которые обрабатываются задачами <b>map</b> полностью параллельно. Среда сортирует выходные данные карт, которые затем вводятся в задачи <b>reduce</b>. Обычно и входные, и выходные данные задания хранятся в файловой системе. Среда занимается планированием задач, их мониторингом и повторно выполняет невыполненные задачи.

Среда MapReduce работает исключительно с парами <ключ, значение>, то есть среда рассматривает входные данные задания как набор пар <ключ, значение> и создает набор пар <ключ, значение> в качестве выходных данных задания, предположительно разных типов.

Типы входных и выходных данных задания MapReduce:
```
(вход) <k1, v1> -> map -> <k2, v2> -> combine -> <k2, v2> -> reduce -> <k3, v3> (выход)
```

Демонстрация для фреймворка MapReduce предназначена для хорошо известной задачи подсчета слов:

![MapReduce](https://www.todaysoftmag.com/images/articles/tsm33/large/a11.png)

# Задание 1: Подсчет количества записей для каждого курса (courseid)

Цель: Научиться использовать MapReduce для подсчета количества записей для каждого курса.

Ход выполнения:
Map: Преобразуйте каждую строку в пару ключ-значение, где ключ — это courseid, а значение — 1.

Reduce: Суммируйте значения для каждого ключа.

In [17]:
class MRCourseCount:
    def __init__(self, input_file, output_file):
        self.input_file = input_file
        self.output_file = output_file

    def mapper(self, line):
        fields = line.strip().split(',')
        if len(fields) > 0:
            courseid = fields[0]
            yield courseid, 1

    def reducer(self, courseid, counts):
        yield courseid, sum(counts)

    def run(self):
        intermediate = {}
        with open(self.input_file, 'r') as f:
            next(f)  # Пропуск заголовка
            for line in f:
                for key, value in self.mapper(line):
                    if key not in intermediate:
                        intermediate[key] = []
                    intermediate[key].append(value)

        results = {}
        for key, values in intermediate.items():
            for result_key, result_value in self.reducer(key, values):
                results[result_key] = result_value

        # Запись результатов в файл
        with open(self.output_file, 'w') as f:
            for courseid, count in results.items():
                f.write(f"{courseid}: {count}\n")


In [19]:
input_file = 'aggrigation_logs_per_week.csv'
output_file = "output.txt"
mr_job = MRCourseCount(input_file, output_file)
mr_job.run()

In [20]:
!ls

'Task 1. Unix.ipynb'			   Task_9.Onto
'Task 2. Terminal.ipynb'		   aggrigation_logs_per_week.csv
'Task 3. Postgre.ipynb'			   big_data_course
'Task 4. PostGre_portal_logs.ipynb'	   dtp201804.csv
'Task 5. Spark.ipynb'			   first1.ipynb
'Task 6. Hadoop.ipynb'			   ontology.owl
'Task 7. Visualization (2).ipynb'	   output.txt
'Task 7. Visualization.ipynb.opdownload'   output_avg_s_all.txt
'Task 8. Map_visualization.ipynb'	   output_low_activity.txt
'Task 9. Ontology.ipynb'		   output_unique_users.txt
 Task_6.MR


# Задание 2: Подсчет среднего значения s_all_avg для каждого курса
С использованием MapReduce вычислить среднее значение  s_all_avg по каждому курсу.

Ход выполнения:
- Map: Преобразуйте каждую строку в пару ключ-значение, где ключ — это courseid, а значение — s_all_avg.

- Reduce: Вычислите среднее значение для каждого курса.

- Сохраните результаты вычисления в отдельный файл

In [21]:
class MRAverageSAll:
    def __init__(self, input_file, output_file):
        self.input_file = input_file
        self.output_file = output_file

    def mapper(self, line):
        fields = line.strip().split(',')
        if len(fields) > 3 and fields[0] and fields[3]:
            courseid = fields[0]
            try:
                s_all_avg = float(fields[3])
                yield courseid, s_all_avg
            except ValueError:
                pass

    def reducer(self, courseid, values):
        total = 0
        count = 0
        for value in values:
            total += value
            count += 1
        if count > 0:
            yield courseid, total / count

    def run(self):
        intermediate = {}
        with open(self.input_file, 'r') as f:
            next(f)
            for line in f:
                for key, value in self.mapper(line):
                    if key not in intermediate:
                        intermediate[key] = []
                    intermediate[key].append(value)

        results = {}
        for key, values in intermediate.items():
            for result_key, result_value in self.reducer(key, values):
                results[result_key] = result_value

        with open(self.output_file, 'w') as f:
            for courseid, avg in results.items():
                f.write(f"{courseid}: {avg}\n")

In [22]:
input_file = 'aggrigation_logs_per_week.csv'
output_file = 'output_avg_s_all.txt'
mr_job = MRAverageSAll(input_file, output_file)
mr_job.run()

# Задание 3: Подсчет количества уникальных пользователей (userid) для каждого курса

Использовать MapReduce для подсчета уникальных значений.

Ход выполнения:
- Map: Преобразуйте каждую строку в пару ключ-значение, где ключ — это courseid, а значение — userid.

- Reduce: Соберите уникальные userid для каждого курса и подсчитайте их количество.

- Сохраните результаты вычисления в отдельный файл

In [23]:
class MRUniqueUsers:
    def __init__(self, input_file, output_file):
        self.input_file = input_file
        self.output_file = output_file

    def mapper(self, line):
        fields = line.strip().split(',')
        if len(fields) > 1 and fields[0] and fields[1]:
            courseid = fields[0]
            userid = fields[1]
            yield courseid, userid

    def reducer(self, courseid, userids):
        unique_users = set(userids)
        yield courseid, len(unique_users)

    def run(self):
        intermediate = {}
        with open(self.input_file, 'r') as f:
            next(f)
            for line in f:
                for key, value in self.mapper(line):
                    if key not in intermediate:
                        intermediate[key] = []
                    intermediate[key].append(value)

        results = {}
        for key, values in intermediate.items():
            for result_key, result_value in self.reducer(key, values):
                results[result_key] = result_value

        with open(self.output_file, 'w') as f:
            for courseid, count in results.items():
                f.write(f"{courseid}: {count}\n")

In [24]:
input_file = 'aggrigation_logs_per_week.csv'
output_file = 'output_unique_users.txt'
mr_job = MRUniqueUsers(input_file, output_file)
mr_job.run()

# Задание 4: Запуск  MapReduce  в терминале

Оформите полученный код из задания 3 в виде отдельного исполняющего файла mr_userid.py

Зайдите в терминал и запустите файл mr_userid.py

Например:
```
python untitled.py aggrigation_logs_per_week.csv
```

Файл  mr_userid.py сохраните в отдельной папке Task_6.MR


================================================


<b>В классической реализации MapReduce (например, в Hadoop) mapper и reducer выполняются как отдельные задачи (отдельные процессы или даже отдельные узлы в кластере). </b>

Преимущества разделения на mapper и reducer:
- Масштабируемость: Mapper и reducer могут выполняться на разных узлах кластера.

- Гибкость: Можно использовать разные языки программирования для mapper и reducer.

- Совместимость с Hadoop: Такой подход легко адаптировать для работы с Hadoop Streaming.

Исправим данное упущение. 

## Шаг 1: Разделение кода на два файла
Файл <b>mapper.py</b>
Этот файл будет содержать логику mapper. Он будет читать входные данные и выдавать промежуточные пары ключ-значение.
Например:
```
import sys

def mapper():
    for line in sys.stdin:
        fields = line.strip().split(',')
        if len(fields) > 1:  # Пропуск пустых строк
            courseid = fields[0]
            userid = fields[1]
            print(f"{courseid}\t{userid}")

if __name__ == '__main__':
    mapper()
```

Файл <b>reducer.py</b>
Этот файл будет содержать логику reducer. Он будет читать промежуточные данные, группировать их по ключам и выполнять агрегацию.
Например:
```
import sys
from itertools import groupby
from operator import itemgetter

def reducer():
    # Чтение данных из stdin и сортировка по ключу (courseid)
    data = [line.strip().split('\t') for line in sys.stdin]
    data.sort(key=itemgetter(0))

    # Группировка по courseid
    for courseid, group in groupby(data, key=itemgetter(0)):
        userids = set(userid for _, userid in group)
        print(f"{courseid}\t{len(userids)}")

if __name__ == '__main__':
    reducer()
```


## Шаг 2: Запуск mapper и reducer через конвейер
Для запуска mapper и reducer как отдельных процессов можно использовать конвейер (pipe) в командной строке. Вот как это сделать:

Запуск mapper и reducer через конвейер:

```
cat /jovyan/aggrigation_logs_per_week.csv | python3 mapper.py | sort | python3 reducer.py > output.txt
```

Здесь:

- cat читает файл и передает его содержимое в mapper.py.

- mapper.py обрабатывает данные и выводит промежуточные пары ключ-значение.

- sort сортирует промежуточные данные по ключу (это важно для правильной работы reducer).

- reducer.py обрабатывает отсортированные данные и записывает результат в output.txt.

Отметим, что при реализации запуска черер конвейер все данные обрабатываются на одной машине.

Преимущества такого запуска:

- Простота: Не требуется Hadoop или кластер.

- Быстрый запуск: Подходит для локальной разработки и тестирования.

- Легкость отладки: Легче отслеживать и исправлять ошибки.

Недостатки:

- Ограниченная масштабируемость: Все данные обрабатываются на одной машине, что не подходит для больших объемов данных.

- Нет отказоустойчивости: Если процесс завершится с ошибкой, все вычисления нужно будет начать заново.

In [25]:
!python Task_6.MR/mr_userid.py aggrigation_logs_per_week.csv

# Задание 5: Выявление студентов с высоким риском отчисления
Найти студентов с низкой активностью, кто может быть в группе риска.
- Mapper:

Для каждой строки данных извлекаем userid и s_all (общая активность студента).

Выдаем пары (userid, s_all).

- Reducer:

Для каждого userid вычисляем среднее значение s_all (средняя активность студента). Вопрос: что не так в данном подходе? Предложите, как исправить упущение (если оно имеется), реализуйте его. 

Сортируем студентов по средней активности и выбираем топ-10 студентов с наименьшей активностью.

Два файла сохранить в папке Task_6.MR и запустить через конвейер. 

#  Запуск через Hadoop Streaming

Hadoop Streaming — это утилита, которая позволяет запускать MapReduce-задачи с использованием любых исполняемых файлов (например, Python-скриптов) в качестве mapper и reducer.

Она работает поверх Hadoop, что позволяет использовать распределенные вычисления на кластере.
Преимущества запуска через Hadoop Streaming:

- Масштабируемость: Hadoop распределяет данные и вычисления по множеству узлов, что позволяет обрабатывать огромные объемы данных.

- Отказоустойчивость: Hadoop автоматически перезапускает задачи, если какой-то узел выходит из строя.

- Интеграция с HDFS: Данные хранятся в HDFS, что обеспечивает высокую производительность и надежность.

Недостатки:

- Требуется настроенный Hadoop-кластер.

- Более сложная настройка и запуск по сравнению с локальным подходом.

Пример запуска:
```
hadoop jar $HADOOP_HOME/share/hadoop/tools/lib/hadoop-streaming-3.2.3.jar \
    -input /jovyan/aggrigation_logs_per_week.csv \
    -output /jovyan/output_stud \
    -mapper "python3 mapper.py" \
    -reducer "python3 reducer.py" \
    -file ./data/big_data_course/Task_6.MR/mapper.py \
    -file ./data/big_data_course/Task_6.MR/reducer.py

```


Обратите внимание, что файл с результатами reducer будет сохранен в директории, которую вы указали в параметре output, причем это выходная директория hdfs, что значит, что вы увидете данный файл не в локальной системе, а именно в hdfs.

Найдем данный файл в hdfs:
```
 hdfs dfs -ls -R /jovyan | grep "output_stud"
```
Пример результата поиска:
```
drwxr-xr-x   - jovyan hadoopusers          0 2025-03-20 14:56 /jovyan/output_stud
-rw-r--r--   1 jovyan hadoopusers          0 2025-03-20 14:56 /jovyan/output_stud/_SUCCESS
-rw-r--r--   1 jovyan hadoopusers       7823 2025-03-20 14:56 /jovyan/output_stud/part-00000
```
Внутри этой директории находятся файлы с результатами, такие как part-00000, part-00001 и т.д. _SUCCESS — пустой файл, который указывает на успешное завершение задачи. part-00000 — файл с результатами.


Чтобы просмотреть содержимое файла с результатами, используйте команду:
```
 hdfs dfs -cat /jovyan/output_stud/part-00000
```

In [26]:
!cat aggrigation_logs_per_week.csv | python3 Task_6.MR/mapper.py | sort | python3 Task_6.MR/reducer.py > output_low_activity.txt

# Задание 6: Запуск MapReduce-задачи через Hadoop Streaming 

Подготовьте команду для запуска MapReduce-задачи через Hadoop Streaming из задания 5.

В терминале проверьте, что задача выполняется успешно.

Откройте полученный файл в hdfs.

Бутьте готовы продемонстриротьва запуск преподавателю. 

In [27]:
!hadoop jar $HADOOP_HOME/share/hadoop/tools/lib/hadoop-streaming-3.2.3.jar \
    -input /jovyan/aggrigation_logs_per_week.csv \
    -output /jovyan/output_low_activity \
    -mapper "python3 mapper.py" \
    -reducer "python3 reducer.py" \
    -file Task_6.MR/mapper.py \
    -file Task_6.MR/reducer.py

2025-06-30 04:44:48,135 WARN streaming.StreamJob: -file option is deprecated, please use generic option -files instead.
packageJobJar: [Task_6.MR/mapper.py, Task_6.MR/reducer.py, /tmp/hadoop-unjar1965665757870867854/] [] /tmp/streamjob3174975863444109571.jar tmpDir=null
2025-06-30 04:44:48,785 INFO client.RMProxy: Connecting to ResourceManager at /0.0.0.0:8032
2025-06-30 04:44:48,932 INFO client.RMProxy: Connecting to ResourceManager at /0.0.0.0:8032
2025-06-30 04:44:49,119 INFO mapreduce.JobResourceUploader: Disabling Erasure Coding for path: /tmp/hadoop-yarn/staging/jovyan/.staging/job_1751258224012_0001
2025-06-30 04:44:50,162 INFO mapred.FileInputFormat: Total input files to process : 1
2025-06-30 04:44:50,213 INFO mapreduce.JobSubmitter: number of splits:2
2025-06-30 04:44:50,702 INFO mapreduce.JobSubmitter: Submitting tokens for job: job_1751258224012_0001
2025-06-30 04:44:50,703 INFO mapreduce.JobSubmitter: Executing with tokens: []
2025-06-30 04:44:50,845 INFO conf.Configuratio

In [29]:
# Проверка наличия выходной директории
!hdfs dfs -ls -R /jovyan | grep "output_low_activity"

drwxr-xr-x   - jovyan hadoopusers          0 2025-06-30 04:45 /jovyan/output_low_activity
-rw-r--r--   1 jovyan hadoopusers          0 2025-06-30 04:45 /jovyan/output_low_activity/_SUCCESS
-rw-r--r--   1 jovyan hadoopusers       7823 2025-06-30 04:45 /jovyan/output_low_activity/part-00000


In [30]:
# Просмотр содержимого результата
!hdfs dfs -cat /jovyan/output_low_activity/part-00000

71262	20
71269	9
71275	9
71279	30
71301	9
71302	6
71323	16
71324	14
71326	11
71335	11
71344	3
71349	14
71352	21
71354	9
71357	30
71384	11
71402	8
71409	11
71411	8
71415	11
71461	14
71467	2
71472	21
71495	67
71496	28
71508	49
71523	14
71535	14
71541	64
71545	45
71547	45
71549	4
71559	12
71571	70
71596	67
71615	31
71619	46
71632	11
71635	20
71646	39
71675	64
71697	45
71702	10
71708	26
71709	46
71710	33
71716	12
71717	28
71720	16
71730	34
71732	37
71733	38
71736	30
71740	9
71746	16
71750	7
71758	7
71776	6
71792	20
71797	19
71801	6
71806	19
71816	13
71820	9
71829	12
71831	11
71852	58
71857	56
71871	13
71880	30
71882	21
71883	27
71884	35
71885	12
71887	8
71892	109
71895	21
71902	11
71903	9
71904	34
71905	11
71906	10
71909	11
71917	38
71925	41
71930	22
71934	47
71963	33
71972	23
71976	47
71978	13
71979	13
71987	19
71995	18
72001	30
72003	17
72007	13
72023	23
72026	13
72028	29
72030	23
72032	8
72053	13
72057	22
72064	5
72073	23
72081	7
72103	18
72107	26
72119	19
72122	36
72126	46
72130	15
721